source/common/upstream/thread_aware_lb_impl.cc

```diff
@@ -0,0 +1,81 @@
+#include "common/upstream/thread_aware_lb_impl.h"
+
+namespace Envoy {
+namespace Upstream {
+
+void ThreadAwareLoadBalancerBase::initialize() {
+  // TODO(mattklein123): In the future, once initialized and the initial LB is built, it would be
+  // better to use a background thread for computing LB updates. This has the substantial benefit
+  // that if the LB computation thread falls behind, host set updates can be trivially collapsed.
+  // I will look into doing this in a follow up. Doing everything using a background thread heavily
+  // complicated initialization as the load balancer would need its own initialized callback. I
+  // think the synchronous/asynchronous split is probably the best option.
+  priority_set_.addMemberUpdateCb(
+      [this](uint32_t, const HostVector&, const HostVector&) -> void { refresh(); });
+
+  refresh();
+}
+
+void ThreadAwareLoadBalancerBase::refresh() {
+  auto per_priority_state = std::make_shared<std::vector<PerPriorityStatePtr>>(
+      priority_set_.hostSetsPerPriority().size());
+  auto per_priority_load = std::make_shared<std::vector<uint32_t>>(per_priority_load_);
+
+  // Note that we only compute global panic on host set refresh. Given that the runtime setting will
+  // rarely change, this is a reasonable compromise to avoid creating extra rings when we only
+  // need to create one per priority level.
+  for (auto& host_set : priority_set_.hostSetsPerPriority()) {
+    const uint32_t priority = host_set->priority();
+    (*per_priority_state)[priority].reset(new PerPriorityState);
```

Comment:
consider making a temp for (*per_priority_state)[priority] for speed (CSE), readability, and debug.